From ac3ccb8bb6ce82c08e80e92b065d85a741100175 Mon Sep 17 00:00:00 2001
From: Jason Lowe <jlowe@yahoo-inc.com>
Date: Wed, 24 May 2017 11:22:42 -0500
Subject: [PATCH 2466/2748] YARN-6615. AmIpFilter drops query parameters on
 redirect. Contributed by Wilfred Spiegelenburg

(cherry picked from commit 8bf1949c0efed700781eb47cf18f9f88443ed506)

Change-Id: I901ef533b189f42aea438cbc33cf66f399bc2976
---
 .../yarn/server/webproxy/amfilter/AmIpFilter.java  |    6 ++++++
 .../server/webproxy/amfilter/TestAmFilter.java     |    8 ++++++++
 2 files changed, 14 insertions(+)

diff --git a/hadoop-yarn-project/hadoop-yarn/hadoop-yarn-server/hadoop-yarn-server-web-proxy/src/main/java/org/apache/hadoop/yarn/server/webproxy/amfilter/AmIpFilter.java b/hadoop-yarn-project/hadoop-yarn/hadoop-yarn-server/hadoop-yarn-server-web-proxy/src/main/java/org/apache/hadoop/yarn/server/webproxy/amfilter/AmIpFilter.java
index 20b79eb..ea9dc7b 100644
--- a/hadoop-yarn-project/hadoop-yarn/hadoop-yarn-server/hadoop-yarn-server-web-proxy/src/main/java/org/apache/hadoop/yarn/server/webproxy/amfilter/AmIpFilter.java
+++ b/hadoop-yarn-project/hadoop-yarn/hadoop-yarn-server/hadoop-yarn-server-web-proxy/src/main/java/org/apache/hadoop/yarn/server/webproxy/amfilter/AmIpFilter.java
@@ -150,6 +150,12 @@ public void doFilter(ServletRequest req, ServletResponse resp,
         insertPoint += PROXY_PATH.length();
         redirect.insert(insertPoint, "/redirect");
       }
+      // add the query parameters on the redirect if there were any
+      String queryString = httpReq.getQueryString();
+      if (queryString != null && !queryString.isEmpty()) {
+        redirect.append("?");
+        redirect.append(queryString);
+      }
 
       ProxyUtils.sendRedirect(httpReq, httpResp, redirect.toString());
       return;
diff --git a/hadoop-yarn-project/hadoop-yarn/hadoop-yarn-server/hadoop-yarn-server-web-proxy/src/test/java/org/apache/hadoop/yarn/server/webproxy/amfilter/TestAmFilter.java b/hadoop-yarn-project/hadoop-yarn/hadoop-yarn-server/hadoop-yarn-server-web-proxy/src/test/java/org/apache/hadoop/yarn/server/webproxy/amfilter/TestAmFilter.java
index 25eae02..fa70a50 100644
--- a/hadoop-yarn-project/hadoop-yarn/hadoop-yarn-server/hadoop-yarn-server-web-proxy/src/test/java/org/apache/hadoop/yarn/server/webproxy/amfilter/TestAmFilter.java
+++ b/hadoop-yarn-project/hadoop-yarn/hadoop-yarn-server/hadoop-yarn-server-web-proxy/src/test/java/org/apache/hadoop/yarn/server/webproxy/amfilter/TestAmFilter.java
@@ -176,6 +176,14 @@ public void doFilter(ServletRequest servletRequest,
     redirect = response.getHeader(ProxyUtils.LOCATION);
     assertEquals("http://bogus/proxy/redirect/application_00_0", redirect);
 
+    // check for query parameters
+    Mockito.when(request.getRequestURI()).thenReturn("/proxy/application_00_0");
+    Mockito.when(request.getQueryString()).thenReturn("id=0");
+    testFilter.doFilter(request, response, chain);
+    assertEquals(HttpURLConnection.HTTP_MOVED_TEMP, response.status);
+    redirect = response.getHeader(ProxyUtils.LOCATION);
+    assertEquals("http://bogus/proxy/redirect/application_00_0?id=0", redirect);
+
     // "127.0.0.1" contains in host list. Without cookie
     Mockito.when(request.getRemoteAddr()).thenReturn("127.0.0.1");
     testFilter.doFilter(request, response, chain);
-- 
1.7.9.5

